### PR TITLE
[MPS] Fix incorrect conv2d output for kernel size >= 256

### LINKED ISSUE #188335: [MPS] Grouped/depthwise Conv2d gives incorrect results for batch_size > 1 (fp32 only; correct at batch_size=1)
### 🐛 Describe the bug

A grouped Conv2d with `groups == in_channels` (kernel spanning the full input height, depthwise-style spatial convolution, as used e.g. in EEGNet's spatial conv layer from the braindecode library) produces incorrect results on MPS whenever batch_size > 1, while batch_size == 1 is correct (max_abs_diff == 0 vs CPU).

While small floating-point differences between backends are expected, this magnitude of error is large enough to make MPS-trained models produce meaningfully different results from CPU-trained ones, which would manifest as accuracy/convergence discrepancies, not just numerical noise.

This reproduces with both the nn.Conv2d module and the functional F.conv2d API, with and without bias, and at every batch size tested (2, 3, 4, 8, 16). It is fully deterministic (identical wrong answer across 5 repeated runs with the same inputs) and does not reproduce in float16, only float32.

### Minimal repro
 
```python
import copy
import torch
import torch.nn as nn
 
torch.manual_seed(0)
device = torch.device("mps")
 
conv_cpu = nn.Conv2d(
    in_channels=8, out_channels=16, kernel_size=(256, 1), groups=8, bias=False
).eval()

conv_mps = copy.deepcopy(conv_cpu).to(device).eval()
 
for batch_size in (1, 2):
    x_cpu = torch.rand(batch_size, 8, 256, 1001)
    x_mps = x_cpu.to(device)
    with torch.no_grad():
        out_cpu = conv_cpu(x_cpu)
        out_mps = conv_mps(x_mps).cpu()
    diff = (out_cpu - out_mps).abs().max().item()
    print(f"batch_size={batch_size}: max_abs_diff={diff:.6f}")
```
 
**Output:**
```
batch_size=1: max_abs_diff=0.000000
batch_size=2: max_abs_diff=1.251600
```

### MPS disagrees with itself, not just with CPU
 
MPS run on a batch of 2 disagrees with MPS run on each of those same two samples individually. The same test run entirely on CPU (as a control) shows exact agreement (`0.000000` for both), confirming this likely is MPS-specific.

```python
x = torch.rand(2, 8, 256, 1001).to(device)
x0, x1 = x[0:1], x[1:2]
 
with torch.no_grad():
    out_batch = conv_mps(x)
    out_cpu = conv_cpu(x.cpu())
    out0 = conv_mps(x0)
    out1 = conv_mps(x1)
 
print((out_batch[0:1] - out0).abs().max().item())
print((out_batch[1:2] - out1).abs().max().item())
print((out_cpu[0:1] - out0.cpu()).abs().max().item())
print((out_cpu[1:2] - out1.cpu()).abs().max().item())
```

### Additional findings
 
- **Deterministic**: identical `max_abs_diff` across 5 repeated runs with the same inputs.
- **Independent of `bias`**: reproduces identically with `bias=True` and `bias=False`.
- **Independent of API surface**: reproduces identically via `F.conv2d` and via the `nn.Conv2d` module.
- **fp32-specific**: does **not** reproduce in `float16` (diff ~0.0002, consistent with normal fp16 precision loss, matching at both batch_size=1  and batch_size=2).
- **Per-sample error pattern is batch-size-dependent and not simply "first sample correct, rest wrong"**. 
```
  batch=2  per-sample diffs: [1.5055, 1.3498]
  batch=3  per-sample diffs: [0.9978, 1.3393, 0.0000]
  batch=4  per-sample diffs: [1.5055, 1.3498, 1.2559, 1.2311]
  batch=8  per-sample diffs: [1.4550, 1.3505, 1.2593, 1.2618, 1.3009, 1.2890, 1.3804, 1.2906]
```

### Versions

PyTorch version: 2.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.5.1 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.1.1.101)
CMake version: version 3.31.6
Libc version: N/A

Python version: 3.14.5 (v3.14.5:5607950ef23, May 10 2026, 07:38:09) [Clang 21.0.0 (clang-2100.0.123.102)] (64-bit runtime)
Python platform: macOS-26.5.1-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] rotary-embedding-torch==0.8.9
[pip3] torch==2.12.0
[pip3] torch-einops-utils==0.1.2
[pip3] torchaudio==2.11.0
[pip3] torchinfo==1.8.0

cc @ezyang @gchanan @kadeng @msaroufim @malfet @aditvenk @snadampal @milpuz01 @aditew01 @nikhil-arm @fadara01 @robert-hardwick @nWEIdia @kulinseth @DenisVieriu97 @jhavukainen @Isalia20

### PR DESCRIPTION
**Pull request by Isalia20**
Fixes #188335

MPSGraph's 2D conv returns wrong results when a kernel spatial dim reaches 256, decompose into sub-256 kernel chunks summed over a padded input so every sub-conv stays correct. While perf won't be ideal, this kernel shape isn't something one will see while training regular computer vision models so this is fine

### CODE REVIEW

**Review by jhavukainen — CHANGES_REQUESTED** [review]
Thanks for the ping. I did some digging on what causes the actual failure and it has to do with the underlying kernel implementation mistakenly dispatching with parameters that'll overrun the available threadgroup memory. I've produced a fix and filed a request for it to get addressed on the OS side.

As to what it means for the pytorch fix: The one you produced almost nails it, but unfortunately the used tg memory scales with both the spatial dim _and_ the batch size in this case. I added some comments on one way to take that into account plus extending the test case a little to expose the remaining problem. 

But I'll admit I didn't yet do much due diligence on if the recursive chunking there is good way to go about it or if there's a clearly better method, just wanted to add it to clarify what the remaining issue is. So feel free to reject the suggestion if you see a better way to handle it.

File: aten/src/ATen/native/mps/operations/Convolution.mm [hunk 1]
```diff
@@ -4,6 +4,7 @@
 #include <ATen/native/mps/OperationUtils.h>
 #include <ATen/ops/_mps_convolution_native.h>
 #include <ATen/ops/_mps_convolution_transpose_native.h>
+#include <ATen/ops/constant_pad_nd.h>
```
Comment (jhavukainen) [inline aten/src/ATen/native/mps/operations/Convolution.mm]:
```suggestion
#include <ATen/ops/constant_pad_nd.h>
#include <ATen/ops/cat.h>
```

File: aten/src/ATen/native/mps/operations/Convolution.mm [hunk 2]
```diff
@@ -152,6 +153,38 @@ static Tensor _mps_convolution_impl(const Tensor& input_t,
                                     IntArrayRef dilation,
                                     int64_t groups,
                                     std::optional<IntArrayRef> input_shape) {
+  // MPSGraph 2D convolution miscomputes the output once a filter dimension reaches 256
+  // since this is not a regular shape that would be seen in regular models, this is fine
+  // i.e. no custom metal kernel needed
+  if (input_t.dim() == 4 && (weight_t.size(2) >= 256 || weight_t.size(3) >= 256)) {
+    constexpr int64_t kChunk = 128;
```
Comment (jhavukainen) [inline aten/src/ATen/native/mps/operations/Convolution.mm]: We need to also partition w.r.t batches since the threadgroup memory the underlying kernel ends up using scales with it as well.

```suggestion
    constexpr int64_t kChunk = 128;
    // The staged threadgroup cube grows with batch as well as filter size, so
    // split the batch and recurse; each chunk then takes the sub-256 path below.
    constexpr int64_t kBatchChunk = 8;
    const int64_t N = input_t.size(0);
    if (N > kBatchChunk) {
      std::vector<Tensor> parts;
      parts.reserve((N + kBatchChunk - 1) / kBatchChunk);
      for (int64_t ns = 0; ns < N; ns += kBatchChunk) {
        parts.push_back(_mps_convolution_impl(input_t.narrow(0, ns, std::min(kBatchChunk, N - ns)), weight_t, bias_opt, padding, stride, dilation, groups, std::nullopt));
      }
      return at::cat(parts, 0);
    }
```

File: test/test_mps.py [hunk 3]
```diff
@@ -2961,6 +2961,34 @@ def helper(input_shape, wt_shape,
             helper((N, C_in * 2, H * 2, W * 2), (C_out * 2, (C_in * 2) // groups,
                    kH + 2, kW + 2), bias_shape=(C_out * 2), groups=groups)
 
+    @parametrize("batch_size", [1, 2])
```
Comment (jhavukainen) [inline test/test_mps.py]: This will help uncover some more failures.

```suggestion
    @parametrize("batch_size", [1, 2, 16, 32])
```

File: aten/src/ATen/native/mps/operations/Convolution.mm (hunk 2 above)
Comment (Isalia20) [reply aten/src/ATen/native/mps/operations/Convolution.mm]: done

File: test/test_mps.py (hunk 3 above)
Comment (Isalia20) [reply test/test_mps.py]: added

File: aten/src/ATen/native/mps/operations/Convolution.mm [hunk 4]
```diff
@@ -152,6 +154,58 @@ static Tensor _mps_convolution_impl(const Tensor& input_t,
...
+      for (int64_t hs = 0; hs < kH; hs += kChunk) {
+        const int64_t ch = std::min(kChunk, kH - hs);
+        const auto rows = padded.narrow(2, hs * dH, (outH - 1) * sH + dH * (ch - 1) + 1);
+        for (int64_t ws = 0; ws < kW; ws += kChunk) {
+          const int64_t cw = std::min(kChunk, kW - ws);
+          const auto x_chunk = rows.narrow(3, ws * dW, (outW - 1) * sW + dW * (cw - 1) + 1).contiguous();
+          const auto w_chunk = weight_t.narrow(2, hs, ch).narrow(3, ws, cw).contiguous();
+          auto part =
+              _mps_convolution_impl(x_chunk, w_chunk, std::nullopt, {0, 0}, stride, dilation, groups, std::nullopt);
+          output_t = output_t.defined() ? output_t.add(part) : std::move(part);
```
Comment (jhavukainen) [inline aten/src/ATen/native/mps/operations/Convolution.mm]: Ok had another look and we might want to consider accumulating in fp16/bf16 cases in fp32 for intermediate results. This will avoid the lower precision rounding error from building up if we have a large number of chunks to add. That being said testing this locally it requires somewhat exotic shapes to be used to end up having large enough number of chunks (>=32) to make it really start showing up. The cost to cast is relatively small compared to the work we already do for the convolution so it should get amortized fairly well.

Since this is already a mechanism to restore correctness to a fairly specific type of shapes hopefully encountered only rarely, I'd argue we should go ahead and go with the higher precision accumulation as well.

```suggestion
          // accumulate in fp32 to avoid aggregating rounding errors from half types
          output_t = output_t.defined() ? output_t.add(part) : part.to(at::kFloat);
```

File: aten/src/ATen/native/mps/operations/Convolution.mm [hunk 5]
```diff
@@ -152,6 +154,58 @@ static Tensor _mps_convolution_impl(const Tensor& input_t,
...
+        }
+      }
+      if (bias_opt && bias_opt->defined()) {
+        output_t = output_t.add(bias_opt->view({1, -1, 1, 1}));
+      }
+      return output_t;
```
Comment (jhavukainen) [inline aten/src/ATen/native/mps/operations/Convolution.mm]: Back to original dtype after addition is done.
```suggestion
      return output_t.to(input_t.scalar_type());
```

File: aten/src/ATen/native/mps/operations/Convolution.mm (hunk 4 above)
Comment (Isalia20) [reply aten/src/ATen/native/mps/operations/Convolution.mm]: makes sense, applied

File: aten/src/ATen/native/mps/operations/Convolution.mm (hunk 5 above)
Comment (Isalia20) [reply aten/src/ATen/native/mps/operations/Convolution.mm]: done

**Review by jhavukainen — APPROVED** [review]

**Review by malfet — APPROVED** [review]
Sure, though I wonder if solution here would be to have a fallback metal kernel implementation, that would work when MPSGraph is broken (similar to Matmul one)

**Comment by malfet** [comment]
@claude review this

File: aten/src/ATen/native/mps/kernels/Convolution.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+#include <c10/metal/common.h>
+
+struct Conv2DParams {
+  int32_t N;
+  int32_t C_in;
+  int32_t C_out;
+  int32_t H;
+  int32_t W;
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/Convolution.h:9]: Nit
```suggestion
  int2 HW;
```

File: aten/src/ATen/native/mps/operations/Convolution.mm [hunk 7]
```diff
@@ -144,6 +152,66 @@ static void fill_conv_desc(MPSGraphConvolution2DOpDescriptor* descriptor_,
...
+  const bool has_bias = bias_opt && bias_opt->defined();
+  const auto bias = has_bias ? bias_opt->contiguous() : Tensor();
+
+  auto output = at::empty(conv_output_size(input.sizes(), weight.sizes(), padding, stride, dilation), input.options());
+  if (output.numel() == 0) {
+    return output;
+  }
+
+  Conv2DParams params;
+  params.N = static_cast<int32_t>(input.size(0));
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/Convolution.mm:175]:
```suggestion
  params.N = c10::checked_convert<int32_t>(input.size(0), "int32_t");
```
